src/Storages/IPartitionStrategy.h

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <Interpreters/ExpressionActions.h>
+#include <Parsers/IAST_fwd.h>
+#include <Storages/KeyDescription.h>
+#include <Processors/Chunk.h>
+
+namespace DB
+{
+
+/*
+ * Class responsible for computing and generating a partition key for object storage.
+ * As of now, there are two possible implementations: hive and wildcard.
+ *
+ * It also offers some helper APIs like `getFormatChunk` and `getFormatHeader`. Required mostly because of `hive` strategy
+ * since the default behavior is not to write partition columns in the files and rely only on the file path.
+ */
+struct IPartitionStrategy
+{
+    struct PartitionExpressionActionsAndColumnName
+    {
+        ExpressionActionsPtr actions;
+        std::string column_name;
+    };
+
+    IPartitionStrategy(KeyDescription partition_key_description_, const Block & sample_block_, ContextPtr context_);
+
+    virtual ~IPartitionStrategy() = default;
+
+    virtual ColumnPtr computePartitionKey(const Chunk & chunk) = 0;
+
+    virtual std::string getPathForRead(const std::string & prefix) = 0;
+    virtual std::string getPathForWrite(const std::string & prefix, const std::string & partition_key) = 0;
+
+    virtual Chunk getFormatChunk(const Chunk & chunk) { return chunk.clone(); }
+
+    virtual Block getFormatHeader() { return sample_block; }
+
+    NamesAndTypesList getPartitionColumns() const;
+    const KeyDescription & getPartitionKeyDescription() const;
+
+protected:
+    const KeyDescription partition_key_description;
+    const Block sample_block;
+    ContextPtr context;
+};
+
+/*
+ * Tries to create a partition strategy given a strategy name.
+ * Performs validation on required arguments by each strategy. Example: Partition strategy hive can not be used without a PARTITION BY expression
+ */
+struct PartitionStrategyFactory
+{
+    enum class StrategyType
+    {
+        NONE, /// The default for data lakes. To keep it backwards compatible, it is promoted to `WILDCARD` in case it is plain object storage
+        WILDCARD,
+        HIVE
+    };
+
+    static std::shared_ptr<IPartitionStrategy> get(
```

Comment:
Where do we use this constructor? I found only usage of the second constructor, but it is easy to miss it

Reply:
The 1st one is used by `StorageFile` and `StorageURL`.

The 2nd one is used by `StorageObjectStorageConfiguration::initPartitionStrategy`.